src/new/wallet.js

```diff
@@ -1334,12 +1337,30 @@ class HathorWallet extends EventEmitter {
 
     newTx.processingStatus = TxHistoryProcessingStatus.PROCESSING;
 
-    // Save the transaction in the storage
+    // Metadata changed MUST be before addTx because we compare the stored tx with the new one.
+    // So overwriting the stored tx before this would make the check invalid.
+    const metadataChanged = await checkTxMetadataChanged(newTx, this.storage);
     await this.storage.addTx(newTx);
-
     await this.scanAddressesToLoad();
+
+    // set state to processing and save current state.
+    const previousState = this.state;
+    this.state = HathorWallet.PROCESSING;
     // Process history to update metadatas
-    await this.storage.processHistory();
+    if (metadataChanged) {
+      // Process the full history because
+      // this tx changed the voided state
+      // XXX in the future we should be able to handle this
+      // voidness alone, without processing everything
+      // but for now it's an isolated case and it's easier
+      await this.storage.processHistory();
+    } else if (isNewTx) {
+      // Process this single transaction.
+      // Handling new metadatas and deleting utxos that are not available anymore
+      await this.storage.processNewTx(newTx);
+    }
+    // restore previous state
+    this.state = previousState;
```

Comment:
_Comment semi-related to line number_

Code looks ok, but should we add tests for this new mechanism?

Reply:
Added the test https://github.com/HathorNetwork/hathor-wallet-lib/pull/810/commits/54c51ceead6667c3c29427d793bcfb71f71dd62f